File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid.Core/Component/ControlTextInputFormat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V4.Content;
using Android.Support.V4.Content.Res;
using Android.Support.V4.Text;
using Android.Text;
using Android.Util;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Core.Class;
using BPMOPMobile.Droid.Core.Common;
using Java.Interop;
using Jp.Wasabeef;

namespace BPMOPMobile.Droid.Core.Component
{
    class ControlTextInputFormat : ControlBase
    {
        private LinearLayout _parentView { get; set; }
        private ViewElement _element { get; set; }
        private RelativeLayout _relaEditor { get; set; }
        private LinearLayout _lnEditor { get; set; } // lớp đè lên rich Editor
        private TextView _tvReadMore { get; set; }

        private EditText _edt;
        private RichEditor _richEditor;
        public override string Value
        {
            set
            {
                var data = value.Trim();
                //this._richEditor.SetHtml(data.ToString());
                //this._edt.TextFormatted = Html.FromHtml(data.ToString());
                this._edt.Text = HtmlCompat.FromHtml(data.ToString(), HtmlCompat.FromHtmlModeLegacy).ToString();

                if (_element.Enable) // Cho chỉnh sửa mới hiện ra Hint
                {
                    if (string.IsNullOrEmpty(data))
                    {
                        this._tvValue.Visibility = ViewStates.Visible;
                        this._tvReadMore.Visibility = ViewStates.Gone;
                        this._edt.Visibility = ViewStates.Gone;

                        this._tvValue.SetTypeface(this._tvValue.Typeface, TypefaceStyle.Italic);
                        this._tvValue.SetTextSize(ComplexUnitType.Sp, 12);
                        this._tvValue.Visibility = ViewStates.Visible;

                        this._tvValue.Text = CmmFunction.GetTitle("TEXT_CHOOSE_CONTENT", "Chọn nội dung...");
                    }
                    else
                    {
                        this._tvValue.Visibility = ViewStates.Gone;
                        this._edt.Visibility = ViewStates.Visible;

                        this._tvReadMore.Post(() =>
                        {
                            // sau khi Set Layout xong sẽ gọi vào hàm này
                            int lines = _edt.LineCount;
                            if (lines > 5)
                                this._tvReadMore.Visibility = ViewStates.Visible;
                            else
                                this._tvReadMore.Visibility = ViewStates.Gone;
                        });
                    }
                }
                else
                {
                    if (!string.IsNullOrEmpty(data))
                    {
                        this._tvValue.Visibility = ViewStates.Gone;
                        this._tvReadMore.Post(() =>
                        {
                            // sau khi Set Layout xong sẽ gọi vào hàm này
                            int lines = _edt.LineCount;
                            if (lines > 5)
                                this._tvReadMore.Visibility = ViewStates.Visible;
                            else
                                this._tvReadMore.Visibility = ViewStates.Gone;
                        });
                    }
                }
            }
            get { return this._richEditor.GetHtml(); }
        }

        public ControlTextInputFormat(Activity _mainAct, LinearLayout _parentView, ViewElement _element) : base(_mainAct)
        {
            this._mainAct = _mainAct;
            this._parentView = _parentView;
            this._element = _element;
            InitializeComponent();
        }
        public override void InitializeComponent()
        {
            base.InitializeComponent();
            _edt = new EditText(_mainAct);
            _relaEditor = new RelativeLayout(_mainAct);
            _lnEditor = new LinearLayout(_mainAct);
            _richEditor = new RichEditor(_mainAct);
            _tvReadMore = new TextView(_mainAct);

            _edt.SetMaxLines(5);
            _edt.Ellipsize = TextUtils.TruncateAt.End;
            _edt.Gravity = GravityFlags.Top;
            _edt.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clBlack)));
            _edt.SetBackgroundColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.transparent)));
            //_edt.Enabled = false;
            _edt.Focusable = false;

            _tvReadMore.SetTextSize(ComplexUnitType.Sp, 12);
            _tvReadMore.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clOrangeFilter)));
            _tvReadMore.SetTypeface(ResourcesCompat.GetFont(_mainAct, Resource.Font.fontarial), TypefaceStyle.Italic);
            _tvReadMore.Ellipsize = TextUtils.TruncateAt.End;
            _tvReadMore.Gravity = GravityFlags.Right;
            _tvReadMore.Visibility = ViewStates.Gone; // Khởi tạo thì ko cần hiện lên, lát check sau
            _tvReadMore.Text = CmmFunction.GetTitle("TEXT_CONTROL_READMORE", "Read more ...");

            _lnEditor.SetBackgroundColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.transparent)));
        }

        public override void InitializeFrameView(LinearLayout frame)
        {
            if (_element.Hidden == true) // Check xem có ẩn view hay không
                return;

            base.InitializeFrameView(frame);

            _lnContent.RemoveView(_lnLine); // Remove line ra để add lại dưới edit text
            _tvValue.Visibility = ViewStates.Gone;

            int _padding = (int)TypedValue.ApplyDimension(ComplexUnitType.Dip, 6, _mainAct.Resources.DisplayMetrics);

            RelativeLayout.LayoutParams _paramsRelaEditor = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MatchParent, RelativeLayout.LayoutParams.WrapContent);
            LinearLayout.LayoutParams _paramsRichEditor = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MatchParent, (int)CmmDroidFunction.ConvertDpToPixel(100, frame.Context));
            LinearLayout.LayoutParams _paramsEditText = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.WrapContent);
            //inearLayout.LayoutParams _paramsLnEditorButton = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MatchParent, (int)CmmDroidFunction.convertDpToPixel(30, frame.Context));
            //LinearLayout.LayoutParams _paramsImgAction = new LinearLayout.LayoutParams((int)CmmDroidFunction.convertDpToPixel(30, frame.Context), (int)CmmDroidFunction.convertDpToPixel(30, frame.Context));

            _edt.LayoutParameters = _paramsEditText;
            _tvReadMore.LayoutParameters = _paramsEditText;

            _relaEditor.LayoutParameters = _paramsRelaEditor;
            _lnEditor.LayoutParameters = _paramsRichEditor;


            _paramsRichEditor.SetMargins(_padding, 0, _padding, 0);
            _tvReadMore.SetPadding(_padding, 0, _padding, 2 * _padding); // đã có padding bottom của Textview Value nên ko cần padding Top nữa
            _richEditor.LayoutParameters = _paramsRichEditor;
            _richEditor.Enabled = false; // disable không cho click

            //_paramsLnEditorButton.SetMargins(_padding, _padding, _padding, _padding);
            //_lnEditorButton.LayoutParameters = _paramsLnEditorButton;
            //_lnEditorButton.Background = ContextCompat.GetDrawable(frame.Context, Resource.Drawable.textcornergray);

            ////_richEditor.Click += HandleTouchDown;
            ////_edt.Click += HandleTouchDown;

            ////_lnEditor.AddView(_lnEditorButton);
            ////_lnEditor.AddView(_edt);

            //_relaEditor.AddView(_richEditor);
            //_relaEditor.AddView(_lnEditor);
            //_relaEditor.AddView(_edt);
            //_relaEditor.AddView(_tvReadMore);

            _lnContent.Click += HandleTouchDown;
            _edt.Click += HandleTouchDown;
            _tvReadMore.Click += HandleTouchDown;
            _lnEditor.Click += HandleTouchDown;

            frame.AddView(_edt);
            frame.AddView(_tvReadMore);
        }
        protected virtual void HandleTouchDown(object sender, EventArgs e)
        {
            if (_parentView != null)
            {
                MinionActionCore.OnElementFormClick(null, new MinionActionCore.ElementFormClick(_element));
            }
        }
        public override void SetEnable()
        {
            base.SetEnable();
            if (_element.Enable)
            {
                _tvValue.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clBlueEnable)));
            }
        }

        public override void SetProprety()
        {
            base.SetProprety();
        }

        public override void SetTitle()
        {
            base.SetTitle();

            _tvTitle.Text = _element.Title;

            if (_element.IsRequire && _element.Enable)
            {
                _tvTitle.Text += " (*)";
                CmmDroidFunction.SetTextViewHighlightControl(_mainAct, _tvTitle);
            }
        }

        public override void SetValue()
        {
            base.SetValue();
            Value = _element.Value;
            if (!String.IsNullOrEmpty(Value))
            {
                _edt.Text = Value;
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/core/component/ControlTextInputFormat.java
package com.vuthao.bpmop.core.component;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Typeface;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.MultipleClickGuard;
import com.vuthao.bpmop.base.custom.editor.RichEditor;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.detail.custom.DetailFunc;

public class ControlTextInputFormat extends ControlBase {
    private final ViewElement element;
    private final LinearLayout parentView;
    private RelativeLayout relaEditor;
    private LinearLayout lnEditor; // lớp đè lên rich Editor
    private TextView tvReadMore;
    private EditText edt;
    private RichEditor richEditor;

    @Override
    public void setValue(String value) {
        String data = value.trim();
        edt.setText(DetailFunc.share.formatHTMLToString(data));

        if (element.isEnable()) {
            if (Functions.isNullOrEmpty(data)) {
                tvValue.setVisibility(View.VISIBLE);
                tvReadMore.setVisibility(View.GONE);
                edt.setVisibility(View.GONE);

                tvValue.setTypeface(super.tvValue.getTypeface(), Typeface.ITALIC);
                tvValue.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
                tvValue.setVisibility(View.VISIBLE);

                tvValue.setText(Functions.share.getTitle("TEXT_CHOOSE_CONTENT", "Chọn nội dung..."));
            } else {
                tvValue.setVisibility(View.GONE);
                edt.setVisibility(View.VISIBLE);

                tvReadMore.post(() -> {
                    int line = edt.getLineCount();
                    if (line > 5) {
                        tvReadMore.setVisibility(View.VISIBLE);
                    } else {
                        tvReadMore.setVisibility(View.GONE);
                    }
                });
            }

        } else {
            if (!Functions.isNullOrEmpty(data)) {
                tvValue.setVisibility(View.GONE);

                tvReadMore.post(() -> {
                    int line = edt.getLineCount();
                    if (line > 5) {
                        tvReadMore.setVisibility(View.VISIBLE);
                    } else {
                        tvReadMore.setVisibility(View.GONE);
                    }
                });
            }
        }
    }

    @Override
    public String getValue() {
        return richEditor.getHtml();
    }

    public ControlTextInputFormat(Activity mainAct, LinearLayout parentView, ViewElement element) {
        super(mainAct);
        this.mainAct = mainAct;
        this.parentView = parentView;
        this.element = element;
        initializeComponent();
    }

    @Override
    public void initializeComponent() {
        super.initializeComponent();

        edt = new EditText(mainAct);
        relaEditor = new RelativeLayout(mainAct);
        lnEditor = new LinearLayout(mainAct);
        richEditor = new RichEditor(mainAct);
        tvReadMore = new TextView(mainAct);

        edt.setMaxLines(3);
        edt.setEllipsize(TextUtils.TruncateAt.END);
        edt.setGravity(Gravity.TOP);
        edt.setTextColor(ContextCompat.getColor(mainAct, R.color.clBlack));
        edt.setBackgroundColor(ContextCompat.getColor(mainAct, R.color.transparent));
        edt.setFocusable(false);

        tvReadMore.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
        tvReadMore.setTextColor(ContextCompat.getColor(mainAct, R.color.clOrangeFilter));
        tvReadMore.setTypeface(ResourcesCompat.getFont(mainAct, R.font.fontarial), Typeface.ITALIC);
        tvReadMore.setEllipsize(TextUtils.TruncateAt.END);
        tvReadMore.setGravity(Gravity.RIGHT);
        tvReadMore.setVisibility(View.GONE);
        tvReadMore.setText(Functions.share.getTitle("TEXT_CONTROL_READMORE", "Read more ..."));
        lnEditor.setBackgroundColor(ContextCompat.getColor(mainAct, R.color.transparent));
    }

    @Override
    public void initializeFrameView(LinearLayout frame) {
        if (element.isHidden()) return;
        super.initializeFrameView(frame);

        lnContent.removeView(lnLine);
        tvValue.setVisibility(View.GONE);

        int padding = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, mainAct.getResources().getDisplayMetrics());
        RelativeLayout.LayoutParams paramsRelaEditor = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
        LinearLayout.LayoutParams paramsRichEditor = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, Functions.share.convertDpToPixel(100, frame.getContext()));
        LinearLayout.LayoutParams paramsEditText = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);

        edt.setLayoutParams(paramsEditText);
        tvReadMore.setLayoutParams(paramsEditText);

        relaEditor.setLayoutParams(paramsRelaEditor);
        lnEditor.setLayoutParams(paramsRichEditor);

        edt.setPadding(padding, 0, padding, 0);
        tvReadMore.setPadding(padding, 0, padding, 2 * padding);
        richEditor.setLayoutParams(paramsRichEditor);
        richEditor.setEnabled(false);

        if (parentView != null) {
            lnContent.setOnClickListener(view -> handleClicks());
            edt.setOnClickListener(view -> handleClicks());
            tvReadMore.setOnClickListener(view -> handleClicks());
            lnEditor.setOnClickListener(view -> handleClicks());
        }

        frame.addView(edt);
        frame.addView(tvReadMore);
    }

    private void handleClicks() {
        new MultipleClickGuard(view -> {
            Intent intent = new Intent();
            intent.setAction(VarsReceiver.FORMCLICK);
            intent.putExtra("element", new Gson().toJson(element));
            BroadcastUtility.send(mainAct, intent);
        }, 2000);
    }

    @Override
    public void setEnable() {
        super.setEnable();
        if (element.isEnable()) {
            tvValue.setTextColor(ContextCompat.getColor(mainAct, R.color.clBlueEnable));
        }
    }

    @Override
    public void setProprety() {
        super.setProprety();
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void setTitle() {
        super.setTitle();

        tvTitle.setText(element.getTitle());
        if (element.isRequire() && element.isEnable()) {
            tvTitle.setText(tvTitle.getText() + " (*)");
            Functions.share.setTVHighlightControl(mainAct, tvTitle);
        }
    }

    @Override
    public void setValue() {
        super.setValue();
        setValue(DetailFunc.share.formatHTMLToString(element.getValue()));
        if (!Functions.isNullOrEmpty(getValue())) {
            edt.setText(getValue());
            edt.setHeight(RelativeLayout.LayoutParams.WRAP_CONTENT);
        }
    }
}
